<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.0//EN" "http://www.w3.org/TR/REC-html40/strict.dtd">

<html><head><title>monarch.api</title><link rel="stylesheet" type="text/css" href="../styles/main.css"><script language=JavaScript src="../javascript/main.js"></script><script language=JavaScript src="../javascript/prettify.js"></script><script language=JavaScript src="../javascript/searchdata.js"></script></head><body class="ContentPage" onLoad="NDOnLoad();prettyPrint();"><script language=JavaScript><!--
if (browserType) {document.write("<div class=" + browserType + ">");if (browserVer) {document.write("<div class=" + browserVer + ">"); }}// --></script>

<!--  Generated by Natural Docs, version 1.5 -->
<!--  http://www.naturaldocs.org  -->

<!-- saved from url=(0026)http://www.naturaldocs.org -->




<div id=Content><div class="CClass"><div class=CTopic id=MainTopic><h1 class=CTitle><a name="monarch.api"></a>monarch.api</h1><div class=CBody><!--START_ND_SUMMARY--><div class=Summary><div class=STitle>Summary</div><div class=SBorder><table border=0 cellspacing=0 cellpadding=0 class=STable><tr class="SMain"><td class=SEntry><a href="#monarch.api" >monarch.api</a></td><td class=SDescription></td></tr><tr class="SGroup SIndent1"><td class=SEntry><a href="#monarch.api.Functions" >Functions</a></td><td class=SDescription></td></tr><tr class="SFunction SIndent2 SMarked"><td class=SEntry><a href="#monarch.api.bbop.monarch.Engine" id=link1 onMouseOver="ShowTip(event, 'tt1', 'link1')" onMouseOut="HideTip('tt1')">bbop.<wbr>monarch.<wbr>Engine</a></td><td class=SDescription>constructor for Engine object</td></tr><tr class="SFunction SIndent2"><td class=SEntry><a href="#monarch.api.fetchDiseaseInfo" id=link2 onMouseOver="ShowTip(event, 'tt2', 'link2')" onMouseOut="HideTip('tt2')">fetchDiseaseInfo</a></td><td class=SDescription>Retrieves JSON block providing info about a disease</td></tr><tr class="SFunction SIndent2 SMarked"><td class=SEntry><a href="#monarch.api.fetchPhenotypeInfo" id=link3 onMouseOver="ShowTip(event, 'tt3', 'link3')" onMouseOut="HideTip('tt3')">fetchPhenotypeInfo</a></td><td class=SDescription>Retrieves JSON block providing info about a phenotype</td></tr><tr class="SFunction SIndent2"><td class=SEntry><a href="#monarch.api.fetchGenotypeInfo" id=link4 onMouseOver="ShowTip(event, 'tt4', 'link4')" onMouseOut="HideTip('tt4')">fetchGenotypeInfo</a></td><td class=SDescription>Retrieves JSON block providing info about a genotype</td></tr><tr class="SFunction SIndent2 SMarked"><td class=SEntry><a href="#monarch.api.fetchGeneInfo" id=link5 onMouseOver="ShowTip(event, 'tt5', 'link5')" onMouseOut="HideTip('tt5')">fetchGeneInfo</a></td><td class=SDescription>Status: PARTIALLY IMPLEMENTED</td></tr><tr class="SFunction SIndent2"><td class=SEntry><a href="#monarch.api.fetchOmimDiseasePhenotypeAsAssocations" id=link6 onMouseOver="ShowTip(event, 'tt6', 'link6')" onMouseOut="HideTip('tt6')">fetchOmimDiseasePhenotypeAsAssocations</a></td><td class=SDescription>Status: IMPLEMENTED</td></tr><tr class="SFunction SIndent2 SMarked"><td class=SEntry><a href="#monarch.api.fetchOmimDiseaseGeneAsAssocations" >fetchOmimDiseaseGeneAsAssocations</a></td><td class=SDescription>Status: IMPLEMENTED</td></tr><tr class="SFunction SIndent2"><td class=SEntry><a href="#monarch.api.fetchOmimGeneAllelePhenotypeAsAssociations" >fetchOmimGeneAllelePhenotypeAsAssociations</a></td><td class=SDescription>Status: DEPRECATED</td></tr><tr class="SFunction SIndent2 SMarked"><td class=SEntry><a href="#monarch.api.makeGenotype" id=link7 onMouseOver="ShowTip(event, 'tt7', 'link7')" onMouseOut="HideTip('tt7')">makeGenotype</a></td><td class=SDescription>Status:  MOSTLY IMPLEMENTED</td></tr><tr class="SFunction SIndent2"><td class=SEntry><a href="#monarch.api.searchByDisease" id=link8 onMouseOver="ShowTip(event, 'tt8', 'link8')" onMouseOut="HideTip('tt8')">searchByDisease</a></td><td class=SDescription>Status:  Partially Implemented</td></tr><tr class="SFunction SIndent2 SMarked"><td class=SEntry><a href="#monarch.api.searchByPhenotypeProfile" id=link9 onMouseOver="ShowTip(event, 'tt9', 'link9')" onMouseOut="HideTip('tt9')">searchByPhenotypeProfile</a></td><td class=SDescription>Returns phenotypically similar entities of a given type.</td></tr><tr class="SFunction SIndent2"><td class=SEntry><a href="#monarch.api.searchByAttributeSet" id=link10 onMouseOver="ShowTip(event, 'tt10', 'link10')" onMouseOut="HideTip('tt10')">searchByAttributeSet</a></td><td class=SDescription>Returns: ordered list of matches</td></tr><tr class="SFunction SIndent2 SMarked"><td class=SEntry><a href="#monarch.api.getAnnotationSufficiencyScore" id=link11 onMouseOver="ShowTip(event, 'tt11', 'link11')" onMouseOut="HideTip('tt11')">getAnnotationSufficiencyScore</a></td><td class=SDescription>Returns: ordered list of matches</td></tr><tr class="SFunction SIndent2"><td class=SEntry><a href="#monarch.api.fetchClassInfo" id=link12 onMouseOver="ShowTip(event, 'tt12', 'link12')" onMouseOut="HideTip('tt12')">fetchClassInfo</a></td><td class=SDescription>Retrieves JSON blob providing info about an ontology class</td></tr><tr class="SFunction SIndent2 SMarked"><td class=SEntry><a href="#monarch.api.keywordSearchOverOntologies" >keywordSearchOverOntologies</a></td><td class=SDescription>Get list of terms matching keywords. </td></tr><tr class="SFunction SIndent2"><td class=SEntry><a href="#monarch.api.searchOverOntologies" id=link13 onMouseOver="ShowTip(event, 'tt13', 'link13')" onMouseOut="HideTip('tt13')">searchOverOntologies</a></td><td class=SDescription>Get list of classes matching keywords</td></tr><tr class="SFunction SIndent2 SMarked"><td class=SEntry><a href="#monarch.api.searchSubstring" id=link14 onMouseOver="ShowTip(event, 'tt14', 'link14')" onMouseOut="HideTip('tt14')">searchSubstring</a></td><td class=SDescription>Services used: NIF vocabulary autocomplete</td></tr><tr class="SFunction SIndent2"><td class=SEntry><a href="#monarch.api.fetchDataFromResource" id=link15 onMouseOver="ShowTip(event, 'tt15', 'link15')" onMouseOut="HideTip('tt15')">fetchDataFromResource</a></td><td class=SDescription>Services used: NIF federation call. </td></tr><tr class="SFunction SIndent2 SMarked"><td class=SEntry><a href="#monarch.api.fetchUrl" id=link16 onMouseOver="ShowTip(event, 'tt16', 'link16')" onMouseOut="HideTip('tt16')">fetchUrl</a></td><td class=SDescription>Generate fetch over HTTP</td></tr><tr class="SFunction SIndent2"><td class=SEntry><a href="#monarch.api.resolveClassId" id=link17 onMouseOver="ShowTip(event, 'tt17', 'link17')" onMouseOut="HideTip('tt17')">resolveClassId</a></td><td class=SDescription>Resolves one of the many identifier styles to a canonical NIF-style class ID</td></tr></table></div></div><!--END_ND_SUMMARY--></div></div></div>

<div class="CGroup"><div class=CTopic><h3 class=CTitle><a name="monarch.api.Functions"></a>Functions</h3></div></div>

<div class="CFunction"><div class=CTopic><h3 class=CTitle><a name="monarch.api.bbop.monarch.Engine"></a>bbop.<wbr>monarch.<wbr>Engine</h3><div class=CBody><blockquote><table border=0 cellspacing=0 cellpadding=0 class="Prototype prettyprint"><tr><td><table border=0 cellspacing=0 cellpadding=0><tr><td class=PBeforeParameters nowrap>bbop.monarch.Engine = function(</td><td class=PParameter nowrap>opts</td><td class=PAfterParameters nowrap>)</td></tr></table></td></tr></table></blockquote><p>constructor for Engine object</p><h4 class=CHeading>Example</h4><p>engine = new bbop.monarch.Engine();</p><h4 class=CHeading>Arguments</h4><p>opts : associative array.&nbsp; Keys: ontology_services_url, federation_services_url</p></div></div></div>

<div class="CFunction"><div class=CTopic><h3 class=CTitle><a name="monarch.api.fetchDiseaseInfo"></a>fetchDiseaseInfo</h3><div class=CBody><blockquote><table border=0 cellspacing=0 cellpadding=0 class="Prototype prettyprint"><tr><td><table border=0 cellspacing=0 cellpadding=0><tr><td class=PBeforeParameters nowrap>bbop.monarch.Engine.prototype.fetchDiseaseInfo = function(</td><td class=PParameter nowrap>id,</td></tr><tr><td></td><td class=PParameter nowrap>opts</td><td class=PAfterParameters nowrap>)</td></tr></table></td></tr></table></blockquote><p>Retrieves JSON block providing info about a disease</p><p>The returned object will be the same as that for fetchClassInfo, enhanced with disease-specific information</p><h4 class=CHeading>Known issues</h4><p>If a DOID is used, similarity scores will not be returned.&nbsp; This should be resolved when the Federation API makes use of equivalence axioms.</p><h4 class=CHeading>Example</h4><p>var diseaseInfo = engine.fetchDiseaseInfo(&ldquo;DOID_1430&rdquo;);</p><h4 class=CHeading>Status</h4><p>status: PARTIALLY IMPLEMENTED</p><h4 class=CHeading>Arguments</h4><p>id : An identifier.&nbsp; One of: IRI string, OBO-style ID or NIF-style ID</p><p>Returns: JSON blob with info about the disease</p></div></div></div>

<div class="CFunction"><div class=CTopic><h3 class=CTitle><a name="monarch.api.fetchPhenotypeInfo"></a>fetchPhenotypeInfo</h3><div class=CBody><blockquote><table border=0 cellspacing=0 cellpadding=0 class="Prototype prettyprint"><tr><td><table border=0 cellspacing=0 cellpadding=0><tr><td class=PBeforeParameters nowrap>bbop.monarch.Engine.prototype.fetchPhenotypeInfo = function(</td><td class=PParameter nowrap>id,</td></tr><tr><td></td><td class=PParameter nowrap>opts</td><td class=PAfterParameters nowrap>)</td></tr></table></td></tr></table></blockquote><p>Retrieves JSON block providing info about a phenotype</p><table border=0 cellspacing=0 cellpadding=0 class=CDescriptionList><tr><td class=CDLEntry>Status: STUB</td><td class=CDLDescription>just diseases.&nbsp; No transitive closure.</td></tr></table><p>Retrieves JSON block providing info about a phenotype</p><p>The returned object will be the same as that for fetchClassInfo, enhanced with phenotype-specific information</p><h4 class=CHeading>Arguments</h4><p>id : An identifier.&nbsp; One of: IRI string, OBO-style ID or NIF-style ID opts : An associative array (EXTENSIBLE, OPTIONAL)</p><p>Returns: JSON blob with info about the phenotype</p></div></div></div>

<div class="CFunction"><div class=CTopic><h3 class=CTitle><a name="monarch.api.fetchGenotypeInfo"></a>fetchGenotypeInfo</h3><div class=CBody><blockquote><table border=0 cellspacing=0 cellpadding=0 class="Prototype prettyprint"><tr><td><table border=0 cellspacing=0 cellpadding=0><tr><td class=PBeforeParameters nowrap>bbop.monarch.Engine.prototype.fetchGenotypeInfo = function(</td><td class=PParameter nowrap>id,</td></tr><tr><td></td><td class=PParameter nowrap>opts</td><td class=PAfterParameters nowrap>)</td></tr></table></td></tr></table></blockquote><p>Retrieves JSON block providing info about a genotype</p><p>The structure follows geno, e.g.</p><blockquote><pre> { id: ZDB-GENO-...?, label:, a&lt;tm1&gt;/a&lt;+&gt;;foo&lt;x&gt;/foo&lt;x&gt;(AB), type: effective_genotype
   has_part : [

     { id: , label:, a&lt;tm1&gt;/a&lt;+&gt;;foo&lt;x&gt;/foo&lt;x&gt;(AB), type: intrinsic_genotype  ## aka SO:genotype
       has_reference_part : {
       id:.., label:AB, type: genomic_background
       }
       has_variant_part : {
         id:.., label:, a&lt;tm1&gt;/a&lt;+&gt;;foo&lt;x&gt;/foo&lt;x&gt;, type: genomic_variation_complement,
         has_variant_part : [
          {
           id:.., label:, a&lt;tm1&gt;/a&lt;+&gt;, type: variation_single_locus_complement,
           has_variant_part : [
            {
             id: ZDB-ALT-, label: a&lt;tm1&gt;, type: variant_locus,
             is_locus_instance_of : {
               id: ZDB-GENE-..., label: a, type: gene_locus,
               has_variant_part : {
                 id:, label: tm1, type: point_mutation
               }
}}]}]}}]}</pre></blockquote><h4 class=CHeading>Arguments</h4><p>id : An identifier.&nbsp; One of: IRI string, OBO-style ID or NIF-style ID opts : An associative array (EXTENSIBLE, OPTIONAL)</p><p>Returns: JSON blob with info about the genotype</p></div></div></div>

<div class="CFunction"><div class=CTopic><h3 class=CTitle><a name="monarch.api.fetchGeneInfo"></a>fetchGeneInfo</h3><div class=CBody><blockquote><table border=0 cellspacing=0 cellpadding=0 class="Prototype prettyprint"><tr><td><table border=0 cellspacing=0 cellpadding=0><tr><td class=PBeforeParameters nowrap>bbop.monarch.Engine.prototype.fetchGeneInfo = function(</td><td class=PParameter nowrap>id,</td></tr><tr><td></td><td class=PParameter nowrap>opts</td><td class=PAfterParameters nowrap>)</td></tr></table></td></tr></table></blockquote><p>Status: PARTIALLY IMPLEMENTED</p><table border=0 cellspacing=0 cellpadding=0 class=CDescriptionList><tr><td class=CDLEntry>TODO</td><td class=CDLDescription>decide whether core gene info should come from ontology or federation</td></tr></table><p>Retrieves JSON block providing info about a gene, currently from MyGene</p><p>The returned object will be the same as that for fetchClassInfo, enhanced with gene-specific information</p><h4 class=CHeading>Arguments</h4><p>id : An identifier.&nbsp; One of: IRI string, OBO-style ID or NIF-style ID</p><p>Returns: JSON blob with info about the gene</p></div></div></div>

<div class="CFunction"><div class=CTopic><h3 class=CTitle><a name="monarch.api.fetchOmimDiseasePhenotypeAsAssocations"></a>fetchOmimDiseasePhenotypeAsAssocations</h3><div class=CBody><blockquote><table border=0 cellspacing=0 cellpadding=0 class="Prototype prettyprint"><tr><td><table border=0 cellspacing=0 cellpadding=0><tr><td class=PBeforeParameters colspan=2>bbop.monarch.Engine.prototype.fetchOmimDiseasePhenotypeAsAssocations = function(</td></tr><tr><td>&nbsp;&nbsp;&nbsp;</td><td class=PParameter nowrap width=100%>id</td></tr><tr><td class=PAfterParameters colspan=2>)</td></tr></table></td></tr></table></blockquote><p>Status: IMPLEMENTED</p><p>Given a query term (e.g. an ID, either disease or phenotype), return an association list object, with structure:</p><p>{ resultCount : NUM, results: [ASSOC1, ...., ASSOCn] }</p><p>Where ASSOC is an associative array representing a disease-phenotype association with the following keys:</p><ul><li>disease : a disease structure</li><li>phenotype : a disease structure</li><li>onset</li><li>frequency</li><li>source</li><li>resource</li></ul><p>Both disease and phenotype structures are associative arrays keyed with id and label.</p><h4 class=CHeading>Stability</h4><ul><li>we may change the JSON object returned to be JSON-LD compliant</li><li>federation query may use ID mapping and OWL equivalence axioms in future</li></ul><h4 class=CHeading>Arguments</h4><ul><li>id : An identifier.&nbsp; One of: IRI string, OBO-style ID or NIF-style ID</li></ul><p>Returns: JSON representing list of D-P associations</p></div></div></div>

<div class="CFunction"><div class=CTopic><h3 class=CTitle><a name="monarch.api.fetchOmimDiseaseGeneAsAssocations"></a>fetchOmimDiseaseGeneAsAssocations</h3><div class=CBody><p>Status: IMPLEMENTED</p><p>Given a query term (e.g. an ID, either disease or phenotype), return an association list object, with structure:</p><p>{ resultCount : NUM, results: [ASSOC1, ...., ASSOCn] }</p><p>Where ASSOC is an associative array representing a disease-gene association.</p><h4 class=CHeading>Stability</h4><ul><li>we may change the JSON object returned to be JSON-LD compliant</li><li>federation query may use ID mapping and OWL equivalence axioms in future</li></ul><h4 class=CHeading>Arguments</h4><ul><li>id : An identifier.&nbsp; One of: IRI string, OBO-style ID or NIF-style ID.&nbsp; E.g.&nbsp; OMIM_105830</li></ul><p>Returns: JSON representing list of D-G associations /</p></div></div></div>

<div class="CFunction"><div class=CTopic><h3 class=CTitle><a name="monarch.api.fetchOmimGeneAllelePhenotypeAsAssociations"></a>fetchOmimGeneAllelePhenotypeAsAssociations</h3><div class=CBody><p>Status: DEPRECATED</p></div></div></div>

<div class="CFunction"><div class=CTopic><h3 class=CTitle><a name="monarch.api.makeGenotype"></a>makeGenotype</h3><div class=CBody><blockquote><table border=0 cellspacing=0 cellpadding=0 class="Prototype prettyprint"><tr><td><table border=0 cellspacing=0 cellpadding=0><tr><td class=PBeforeParameters nowrap>bbop.monarch.Engine.prototype.makeGenotype = function(</td><td class=PParameter nowrap>r</td><td class=PAfterParameters nowrap>)</td></tr></table></td></tr></table></blockquote><p>Status:  MOSTLY IMPLEMENTED</p><p>This is meant as an internal function to take a row &reg; from NIF geno or pheno tables, and extract out the parts pertaining to a genotype, and build the JSON genotype object, following the GENO model.&nbsp; This will probably go away once we are able to get this data from D2R transformed datastore What is presently missing is the aggregation of genotype parts.&nbsp;  So, because the source data is denormalized, one genotype can have multiple rows, and will currently create multiple JSON objects with only partially-overlapping contents.&nbsp;  This will be confusing.</p></div></div></div>

<div class="CFunction"><div class=CTopic><h3 class=CTitle><a name="monarch.api.searchByDisease"></a>searchByDisease</h3><div class=CBody><blockquote><table border=0 cellspacing=0 cellpadding=0 class="Prototype prettyprint"><tr><td><table border=0 cellspacing=0 cellpadding=0><tr><td class=PBeforeParameters nowrap>bbop.monarch.Engine.prototype.searchByDisease = function(</td><td class=PParameter nowrap>id,</td></tr><tr><td></td><td class=PParameter nowrap>target_species,</td></tr><tr><td></td><td class=PParameter nowrap>cutoff,</td></tr><tr><td></td><td class=PParameter nowrap>metric</td><td class=PAfterParameters nowrap>)</td></tr></table></td></tr></table></blockquote><p>Status:  Partially Implemented</p><p>This function will take some kind of identifier (disease, genotype), and find those &ldquo;things&rdquo; that are similar.&nbsp; First implementation will take a disease or genotype identifier, and find the top &ldquo;cutoff&rdquo; things (either diseases or animal models) based on the owlsim precomputed piepline.&nbsp;  These are loaded statically in the NIF/Monarch tables, with a resource identification map to locate the correct precomputed table in NIF.&nbsp; Defaults are set for cutoff(25) and metric (BMAasymIC), but the query id and target_species should be supplied in order to know what table needs to be fetched This currently does an iterative lookup in order to get all the labels.&nbsp;  Note sure if that is the best strategy.&nbsp; This wrapping function could easily have precomputed or on-the-fly functions added or swapped.</p><p>TODO: this can be extended to apply other kinds of filters to limit the data considered, or only specific classes or whatever.&nbsp;  Lots of possiblities for the future.&nbsp; TODO: this should really have a target_type filter, and the selection of the precomputed dataset should be based on the combination of source_taxid, target_taxid, a_type, and b_type.</p></div></div></div>

<div class="CFunction"><div class=CTopic><h3 class=CTitle><a name="monarch.api.searchByPhenotypeProfile"></a>searchByPhenotypeProfile</h3><div class=CBody><blockquote><table border=0 cellspacing=0 cellpadding=0 class="Prototype prettyprint"><tr><td><table border=0 cellspacing=0 cellpadding=0><tr><td class=PBeforeParameters colspan=2>bbop.monarch.Engine.prototype.searchByPhenotypeProfile = function(</td></tr><tr><td>&nbsp;&nbsp;&nbsp;</td><td class=PParameter nowrap width=100%>query,</td></tr><tr><td>&nbsp;&nbsp;&nbsp;</td><td class=PParameter nowrap width=100%>target_species,</td></tr><tr><td>&nbsp;&nbsp;&nbsp;</td><td class=PParameter nowrap width=100%>target_type,</td></tr><tr><td>&nbsp;&nbsp;&nbsp;</td><td class=PParameter nowrap width=100%>cutoff,</td></tr><tr><td>&nbsp;&nbsp;&nbsp;</td><td class=PParameter nowrap width=100%>metric</td></tr><tr><td class=PAfterParameters colspan=2>)</td></tr></table></td></tr></table></blockquote><p>Returns phenotypically similar entities of a given type.</p><p>This is currently hardcoded to types per species, but will be expanded in the future to be more customizable.</p><p>Status: PARTIALLY IMPLEMENTED</p><h4 class=CHeading>Arguments</h4><ul><li>query : a list of phenotype identifiers</li><li>target_species : numeric fragment of NCBITaxon identifier</li><li>target_type : genotype | disease | gene | variation (NOT IMPLEMENTED)</li><li>cutoff :</li><li>metric :</li></ul></div></div></div>

<div class="CFunction"><div class=CTopic><h3 class=CTitle><a name="monarch.api.searchByAttributeSet"></a>searchByAttributeSet</h3><div class=CBody><blockquote><table border=0 cellspacing=0 cellpadding=0 class="Prototype prettyprint"><tr><td><table border=0 cellspacing=0 cellpadding=0><tr><td class=PBeforeParameters nowrap>bbop.monarch.Engine.prototype.searchByAttributeSet = function(</td><td class=PParameter nowrap>atts,</td></tr><tr><td></td><td class=PParameter nowrap>tf,</td></tr><tr><td></td><td class=PParameter nowrap>limit</td><td class=PAfterParameters nowrap>)</td></tr></table></td></tr></table></blockquote><p>Returns: ordered list of matches</p><h4 class=CHeading>Arguments</h4><ul><li>atts : list of class identifiers</li><li>target_filter: An ID space (string) [e.g.&nbsp; ZFIN] OR a query object</li><li>limit : number of results to return.</li></ul><p>Query objects are only partially supported.&nbsp; An ID space can be used in place of an object, but this usage may be deprecated in future.</p><h4 class=CHeading>Keys</h4><ul><li>target: an ID space</li><li>species: numeric fragment of NCBITaxon ID or IRI</li><li>type: one of gene | genotype | disease (NOT IMPLEMENTED)</li></ul></div></div></div>

<div class="CFunction"><div class=CTopic><h3 class=CTitle><a name="monarch.api.getAnnotationSufficiencyScore"></a>getAnnotationSufficiencyScore</h3><div class=CBody><blockquote><table border=0 cellspacing=0 cellpadding=0 class="Prototype prettyprint"><tr><td><table border=0 cellspacing=0 cellpadding=0><tr><td class=PBeforeParameters nowrap>bbop.monarch.Engine.prototype.getAnnotationSufficiencyScore = function(</td><td class=PParameter nowrap>atts</td><td class=PAfterParameters nowrap>)</td></tr></table></td></tr></table></blockquote><p>Returns: ordered list of matches</p><h4 class=CHeading>Arguments</h4><ul><li>atts : list of class identifiers</li><li>target: (optional) e.g.&nbsp; MGI</li><li>limit : number of results to return</li></ul></div></div></div>

<div class="CFunction"><div class=CTopic><h3 class=CTitle><a name="monarch.api.fetchClassInfo"></a>fetchClassInfo</h3><div class=CBody><blockquote><table border=0 cellspacing=0 cellpadding=0 class="Prototype prettyprint"><tr><td><table border=0 cellspacing=0 cellpadding=0><tr><td class=PBeforeParameters nowrap>bbop.monarch.Engine.prototype.fetchClassInfo = function(</td><td class=PParameter nowrap>id,</td></tr><tr><td></td><td class=PParameter nowrap>opts</td><td class=PAfterParameters nowrap>)</td></tr></table></td></tr></table></blockquote><p>Retrieves JSON blob providing info about an ontology class</p><h4 class=CHeading>Data structures</h4><ul><li>Class = { id: ID, label : LABEL, relationships: [Relationship*] OPTIONAL }</li><li>Relationship = { subject: OBJ, property : OBJ, object : OBJ }</li><li>Obj = { id : ID, label : LABEL }</li></ul><h4 class=CHeading>Services used</h4><ul><li>OntoQuest</li></ul><h4 class=CHeading>Options</h4><ul><li>level : See OntoQuest docs.&nbsp; If set, a list of relationships will form a graph with distance=level focused on class</li></ul><h4 class=CHeading>Implementation notes</h4><ul><li>makes use of OntoQuest REST API which returns XML</li></ul><h4 class=CHeading>Arguments</h4><ul><li>id : An identifier.&nbsp; One of: IRI string, OBO-style ID or NIF-style ID</li><li>opts : Dictionary parameter</li></ul><p>Returns: Class struct as JSON</p></div></div></div>

<div class="CFunction"><div class=CTopic><h3 class=CTitle><a name="monarch.api.keywordSearchOverOntologies"></a>keywordSearchOverOntologies</h3><div class=CBody><p>Get list of terms matching keywords.&nbsp; Here a &lsquo;term&rsquo; is just a string (TODO: find if there is a way to get OQ return ID+label pairs)</p><p>TODO: all ancestors.&nbsp; E.g.&nbsp; <a href="http://nif-services.neuinfo.org/ontoquest/keyword/glial+c/anc/terms/cell;process" class=LURL target=_top>http://nif-services.neuinfo.org<wbr>/ontoquest<wbr>/keyword<wbr>/glial+c<wbr>/anc<wbr>/terms<wbr>/cell;process</a></p><p>Can be used for auto-complete</p><h4 class=CHeading>Arguments</h4><ul><li>term : keyword</li><li>opts : Dictionary parameter</li></ul><p>Returns: Class struct as JSON</p></div></div></div>

<div class="CFunction"><div class=CTopic><h3 class=CTitle><a name="monarch.api.searchOverOntologies"></a>searchOverOntologies</h3><div class=CBody><blockquote><table border=0 cellspacing=0 cellpadding=0 class="Prototype prettyprint"><tr><td><table border=0 cellspacing=0 cellpadding=0><tr><td class=PBeforeParameters nowrap>bbop.monarch.Engine.prototype.searchOverOntologies = function(</td><td class=PParameter nowrap>term,</td></tr><tr><td></td><td class=PParameter nowrap>opts</td><td class=PAfterParameters nowrap>)</td></tr></table></td></tr></table></blockquote><p>Get list of classes matching keywords</p><p>TODO: all ancestors.&nbsp; E.g.&nbsp; <a href="http://nif-services.neuinfo.org/ontoquest/keyword/glial+c/anc/terms/cell;process" class=LURL target=_top>http://nif-services.neuinfo.org<wbr>/ontoquest<wbr>/keyword<wbr>/glial+c<wbr>/anc<wbr>/terms<wbr>/cell;process</a></p><h4 class=CHeading>Arguments</h4><ul><li>term : keyword</li><li>opts : Dictionary parameter</li></ul><p>Returns: Class struct as JSON</p></div></div></div>

<div class="CFunction"><div class=CTopic><h3 class=CTitle><a name="monarch.api.searchSubstring"></a>searchSubstring</h3><div class=CBody><blockquote><table border=0 cellspacing=0 cellpadding=0 class="Prototype prettyprint"><tr><td><table border=0 cellspacing=0 cellpadding=0><tr><td class=PBeforeParameters nowrap>bbop.monarch.Engine.prototype.searchSubstring = function(</td><td class=PParameter nowrap>term</td><td class=PAfterParameters nowrap>)</td></tr></table></td></tr></table></blockquote><p>Services used: NIF vocabulary autocomplete</p><h4 class=CHeading>Arguments</h4><p>term : search term Returns: JSON structure - as per NIF vocabular services</p></div></div></div>

<div class="CFunction"><div class=CTopic><h3 class=CTitle><a name="monarch.api.fetchDataFromResource"></a>fetchDataFromResource</h3><div class=CBody><blockquote><table border=0 cellspacing=0 cellpadding=0 class="Prototype prettyprint"><tr><td><table border=0 cellspacing=0 cellpadding=0><tr><td class=PBeforeParameters nowrap>bbop.monarch.Engine.prototype.fetchDataFromResource = function(</td><td class=PParameter nowrap>id,</td></tr><tr><td></td><td class=PParameter nowrap>resource_id,</td></tr><tr><td></td><td class=PParameter nowrap>trFunction,</td></tr><tr><td></td><td class=PParameter nowrap>projectCols,</td></tr><tr><td></td><td class=PParameter nowrap>filters,</td></tr><tr><td></td><td class=PParameter nowrap>limit,</td></tr><tr><td></td><td class=PParameter nowrap>sortField,</td></tr><tr><td></td><td class=PParameter nowrap>opts</td><td class=PAfterParameters nowrap>)</td></tr></table></td></tr></table></blockquote><p>Services used: NIF federation call.&nbsp; See: <a href="http://beta.neuinfo.org/services/resource_FederationService.html#path__federation_data_-nifId-.html" class=LURL target=_top>http://beta.neuinfo.org<wbr>/services<wbr>/resource_FederationService.html#path__federation_data_-nifId-.html</a></p><p>This currently relies on OntoQuest magic to ensure that the input ontology class ID is expanded to all relevant <a href="https://support.crbs.ucsd.edu/browse/LAMHDI-140" class=LURL target=_top>https://support.crbs.ucsd.edu/browse/LAMHDI-140</a></p><h4 class=CHeading>Arguments</h4><p>q : query.&nbsp; Arbitrary term or NIF ID resource_id : E.g. nlx_151835-1 trFunction : a function to be applied to each result object which returns a transformed object project    : a list of columns to project (select) filters    : a list of filters to be applied at fetching time Returns: JSON structure { resultCount : NUM, results: [ TRANSFORMED-OBJECTS ] }</p></div></div></div>

<div class="CFunction"><div class=CTopic><h3 class=CTitle><a name="monarch.api.fetchUrl"></a>fetchUrl</h3><div class=CBody><blockquote><table border=0 cellspacing=0 cellpadding=0 class="Prototype prettyprint"><tr><td><table border=0 cellspacing=0 cellpadding=0><tr><td class=PBeforeParameters nowrap>bbop.monarch.Engine.prototype.fetchUrl = function(</td><td class=PParameter nowrap>url,</td></tr><tr><td></td><td class=PParameter nowrap>params,</td></tr><tr><td></td><td class=PParameter nowrap>method</td><td class=PAfterParameters nowrap>)</td></tr></table></td></tr></table></blockquote><p>Generate fetch over HTTP</p><p>In future this will abstract over base implementation: rhino vs jquery</p><h4 class=CHeading>Arguments</h4><p>url : string params :   string OR list OR dict</p><table border=0 cellspacing=0 cellpadding=0 class=CDescriptionList><tr><td class=CDLEntry>Returns: string</td><td class=CDLDescription>may be JSON, XML, who knows</td></tr></table></div></div></div>

<div class="CFunction"><div class=CTopic><h3 class=CTitle><a name="monarch.api.resolveClassId"></a>resolveClassId</h3><div class=CBody><blockquote><table border=0 cellspacing=0 cellpadding=0 class="Prototype prettyprint"><tr><td><table border=0 cellspacing=0 cellpadding=0><tr><td class=PBeforeParameters nowrap>bbop.monarch.Engine.prototype.resolveClassId = function(</td><td class=PParameter nowrap>id</td><td class=PAfterParameters nowrap>)</td></tr></table></td></tr></table></blockquote><p>Resolves one of the many identifier styles to a canonical NIF-style class ID</p><p>Example: resolveClassId(&ldquo;OMIM:123456&rdquo;) returns &ldquo;OMIM_123456&rdquo;</p><h4 class=CHeading>Arguments</h4><p>id : string</p><p>Returns: NIF-style ID</p></div></div></div>

</div><!--Content-->


<div id=Footer><a href="http://www.naturaldocs.org">Generated by Natural Docs</a></div><!--Footer-->


<div id=Menu><div class=MEntry><div class=MFile id=MSelected>api.js</div></div><div class=MEntry><div class=MFile><a href="web/webapp-js.html">webapp</a></div></div><div class=MEntry><div class=MGroup><a href="javascript:ToggleMenu('MGroupContent1')">Index</a><div class=MGroupContent id=MGroupContent1><div class=MEntry><div class=MIndex><a href="../index/General.html">Everything</a></div></div><div class=MEntry><div class=MIndex><a href="../index/Classes.html">Classes</a></div></div><div class=MEntry><div class=MIndex><a href="../index/Functions.html">Functions</a></div></div></div></div></div><script type="text/javascript"><!--
var searchPanel = new SearchPanel("searchPanel", "HTML", "../search");
--></script><div id=MSearchPanel class=MSearchPanelInactive><input type=text id=MSearchField value=Search onFocus="searchPanel.OnSearchFieldFocus(true)" onBlur="searchPanel.OnSearchFieldFocus(false)" onKeyUp="searchPanel.OnSearchFieldChange()"><select id=MSearchType onFocus="searchPanel.OnSearchTypeFocus(true)" onBlur="searchPanel.OnSearchTypeFocus(false)" onChange="searchPanel.OnSearchTypeChange()"><option  id=MSearchEverything selected value="General">Everything</option><option value="Classes">Classes</option><option value="Functions">Functions</option></select></div></div><!--Menu-->



<!--START_ND_TOOLTIPS-->
<div class=CToolTip id="tt1"><div class=CFunction><blockquote><table border=0 cellspacing=0 cellpadding=0 class="Prototype prettyprint"><tr><td><table border=0 cellspacing=0 cellpadding=0><tr><td class=PBeforeParameters nowrap>bbop.monarch.Engine = function(</td><td class=PParameter nowrap>opts</td><td class=PAfterParameters nowrap>)</td></tr></table></td></tr></table></blockquote>constructor for Engine object</div></div><div class=CToolTip id="tt2"><div class=CFunction><blockquote><table border=0 cellspacing=0 cellpadding=0 class="Prototype prettyprint"><tr><td><table border=0 cellspacing=0 cellpadding=0><tr><td class=PBeforeParameters nowrap>bbop.monarch.Engine.prototype.fetchDiseaseInfo = function(</td><td class=PParameter nowrap>id,</td></tr><tr><td></td><td class=PParameter nowrap>opts</td><td class=PAfterParameters nowrap>)</td></tr></table></td></tr></table></blockquote>Retrieves JSON block providing info about a disease</div></div><div class=CToolTip id="tt3"><div class=CFunction><blockquote><table border=0 cellspacing=0 cellpadding=0 class="Prototype prettyprint"><tr><td><table border=0 cellspacing=0 cellpadding=0><tr><td class=PBeforeParameters nowrap>bbop.monarch.Engine.prototype.fetchPhenotypeInfo = function(</td><td class=PParameter nowrap>id,</td></tr><tr><td></td><td class=PParameter nowrap>opts</td><td class=PAfterParameters nowrap>)</td></tr></table></td></tr></table></blockquote>Retrieves JSON block providing info about a phenotype</div></div><div class=CToolTip id="tt4"><div class=CFunction><blockquote><table border=0 cellspacing=0 cellpadding=0 class="Prototype prettyprint"><tr><td><table border=0 cellspacing=0 cellpadding=0><tr><td class=PBeforeParameters nowrap>bbop.monarch.Engine.prototype.fetchGenotypeInfo = function(</td><td class=PParameter nowrap>id,</td></tr><tr><td></td><td class=PParameter nowrap>opts</td><td class=PAfterParameters nowrap>)</td></tr></table></td></tr></table></blockquote>Retrieves JSON block providing info about a genotype</div></div><div class=CToolTip id="tt5"><div class=CFunction><blockquote><table border=0 cellspacing=0 cellpadding=0 class="Prototype prettyprint"><tr><td><table border=0 cellspacing=0 cellpadding=0><tr><td class=PBeforeParameters nowrap>bbop.monarch.Engine.prototype.fetchGeneInfo = function(</td><td class=PParameter nowrap>id,</td></tr><tr><td></td><td class=PParameter nowrap>opts</td><td class=PAfterParameters nowrap>)</td></tr></table></td></tr></table></blockquote>Status: PARTIALLY IMPLEMENTED</div></div><div class=CToolTip id="tt6"><div class=CFunction><blockquote><table border=0 cellspacing=0 cellpadding=0 class="Prototype prettyprint"><tr><td><table border=0 cellspacing=0 cellpadding=0><tr><td class=PBeforeParameters colspan=2>bbop.monarch.Engine.prototype.fetchOmimDiseasePhenotypeAsAssocations = function(</td></tr><tr><td>&nbsp;&nbsp;&nbsp;</td><td class=PParameter nowrap width=100%>id</td></tr><tr><td class=PAfterParameters colspan=2>)</td></tr></table></td></tr></table></blockquote>Status: IMPLEMENTED</div></div><div class=CToolTip id="tt7"><div class=CFunction><blockquote><table border=0 cellspacing=0 cellpadding=0 class="Prototype prettyprint"><tr><td><table border=0 cellspacing=0 cellpadding=0><tr><td class=PBeforeParameters nowrap>bbop.monarch.Engine.prototype.makeGenotype = function(</td><td class=PParameter nowrap>r</td><td class=PAfterParameters nowrap>)</td></tr></table></td></tr></table></blockquote>Status:  MOSTLY IMPLEMENTED</div></div><div class=CToolTip id="tt8"><div class=CFunction><blockquote><table border=0 cellspacing=0 cellpadding=0 class="Prototype prettyprint"><tr><td><table border=0 cellspacing=0 cellpadding=0><tr><td class=PBeforeParameters nowrap>bbop.monarch.Engine.prototype.searchByDisease = function(</td><td class=PParameter nowrap>id,</td></tr><tr><td></td><td class=PParameter nowrap>target_species,</td></tr><tr><td></td><td class=PParameter nowrap>cutoff,</td></tr><tr><td></td><td class=PParameter nowrap>metric</td><td class=PAfterParameters nowrap>)</td></tr></table></td></tr></table></blockquote>Status:  Partially Implemented</div></div><div class=CToolTip id="tt9"><div class=CFunction><blockquote><table border=0 cellspacing=0 cellpadding=0 class="Prototype prettyprint"><tr><td><table border=0 cellspacing=0 cellpadding=0><tr><td class=PBeforeParameters colspan=2>bbop.monarch.Engine.prototype.searchByPhenotypeProfile = function(</td></tr><tr><td>&nbsp;&nbsp;&nbsp;</td><td class=PParameter nowrap width=100%>query,</td></tr><tr><td>&nbsp;&nbsp;&nbsp;</td><td class=PParameter nowrap width=100%>target_species,</td></tr><tr><td>&nbsp;&nbsp;&nbsp;</td><td class=PParameter nowrap width=100%>target_type,</td></tr><tr><td>&nbsp;&nbsp;&nbsp;</td><td class=PParameter nowrap width=100%>cutoff,</td></tr><tr><td>&nbsp;&nbsp;&nbsp;</td><td class=PParameter nowrap width=100%>metric</td></tr><tr><td class=PAfterParameters colspan=2>)</td></tr></table></td></tr></table></blockquote>Returns phenotypically similar entities of a given type.</div></div><div class=CToolTip id="tt10"><div class=CFunction><blockquote><table border=0 cellspacing=0 cellpadding=0 class="Prototype prettyprint"><tr><td><table border=0 cellspacing=0 cellpadding=0><tr><td class=PBeforeParameters nowrap>bbop.monarch.Engine.prototype.searchByAttributeSet = function(</td><td class=PParameter nowrap>atts,</td></tr><tr><td></td><td class=PParameter nowrap>tf,</td></tr><tr><td></td><td class=PParameter nowrap>limit</td><td class=PAfterParameters nowrap>)</td></tr></table></td></tr></table></blockquote>Returns: ordered list of matches</div></div><div class=CToolTip id="tt11"><div class=CFunction><blockquote><table border=0 cellspacing=0 cellpadding=0 class="Prototype prettyprint"><tr><td><table border=0 cellspacing=0 cellpadding=0><tr><td class=PBeforeParameters nowrap>bbop.monarch.Engine.prototype.getAnnotationSufficiencyScore = function(</td><td class=PParameter nowrap>atts</td><td class=PAfterParameters nowrap>)</td></tr></table></td></tr></table></blockquote>Returns: ordered list of matches</div></div><div class=CToolTip id="tt12"><div class=CFunction><blockquote><table border=0 cellspacing=0 cellpadding=0 class="Prototype prettyprint"><tr><td><table border=0 cellspacing=0 cellpadding=0><tr><td class=PBeforeParameters nowrap>bbop.monarch.Engine.prototype.fetchClassInfo = function(</td><td class=PParameter nowrap>id,</td></tr><tr><td></td><td class=PParameter nowrap>opts</td><td class=PAfterParameters nowrap>)</td></tr></table></td></tr></table></blockquote>Retrieves JSON blob providing info about an ontology class</div></div><div class=CToolTip id="tt13"><div class=CFunction><blockquote><table border=0 cellspacing=0 cellpadding=0 class="Prototype prettyprint"><tr><td><table border=0 cellspacing=0 cellpadding=0><tr><td class=PBeforeParameters nowrap>bbop.monarch.Engine.prototype.searchOverOntologies = function(</td><td class=PParameter nowrap>term,</td></tr><tr><td></td><td class=PParameter nowrap>opts</td><td class=PAfterParameters nowrap>)</td></tr></table></td></tr></table></blockquote>Get list of classes matching keywords</div></div><div class=CToolTip id="tt14"><div class=CFunction><blockquote><table border=0 cellspacing=0 cellpadding=0 class="Prototype prettyprint"><tr><td><table border=0 cellspacing=0 cellpadding=0><tr><td class=PBeforeParameters nowrap>bbop.monarch.Engine.prototype.searchSubstring = function(</td><td class=PParameter nowrap>term</td><td class=PAfterParameters nowrap>)</td></tr></table></td></tr></table></blockquote>Services used: NIF vocabulary autocomplete</div></div><div class=CToolTip id="tt15"><div class=CFunction><blockquote><table border=0 cellspacing=0 cellpadding=0 class="Prototype prettyprint"><tr><td><table border=0 cellspacing=0 cellpadding=0><tr><td class=PBeforeParameters nowrap>bbop.monarch.Engine.prototype.fetchDataFromResource = function(</td><td class=PParameter nowrap>id,</td></tr><tr><td></td><td class=PParameter nowrap>resource_id,</td></tr><tr><td></td><td class=PParameter nowrap>trFunction,</td></tr><tr><td></td><td class=PParameter nowrap>projectCols,</td></tr><tr><td></td><td class=PParameter nowrap>filters,</td></tr><tr><td></td><td class=PParameter nowrap>limit,</td></tr><tr><td></td><td class=PParameter nowrap>sortField,</td></tr><tr><td></td><td class=PParameter nowrap>opts</td><td class=PAfterParameters nowrap>)</td></tr></table></td></tr></table></blockquote>Services used: NIF federation call. </div></div><div class=CToolTip id="tt16"><div class=CFunction><blockquote><table border=0 cellspacing=0 cellpadding=0 class="Prototype prettyprint"><tr><td><table border=0 cellspacing=0 cellpadding=0><tr><td class=PBeforeParameters nowrap>bbop.monarch.Engine.prototype.fetchUrl = function(</td><td class=PParameter nowrap>url,</td></tr><tr><td></td><td class=PParameter nowrap>params,</td></tr><tr><td></td><td class=PParameter nowrap>method</td><td class=PAfterParameters nowrap>)</td></tr></table></td></tr></table></blockquote>Generate fetch over HTTP</div></div><div class=CToolTip id="tt17"><div class=CFunction><blockquote><table border=0 cellspacing=0 cellpadding=0 class="Prototype prettyprint"><tr><td><table border=0 cellspacing=0 cellpadding=0><tr><td class=PBeforeParameters nowrap>bbop.monarch.Engine.prototype.resolveClassId = function(</td><td class=PParameter nowrap>id</td><td class=PAfterParameters nowrap>)</td></tr></table></td></tr></table></blockquote>Resolves one of the many identifier styles to a canonical NIF-style class ID</div></div><!--END_ND_TOOLTIPS-->




<div id=MSearchResultsWindow><iframe src="" frameborder=0 name=MSearchResults id=MSearchResults></iframe><a href="javascript:searchPanel.CloseResultsWindow()" id=MSearchResultsWindowClose>Close</a></div>


<script language=JavaScript><!--
if (browserType) {if (browserVer) {document.write("</div>"); }document.write("</div>");}// --></script></body></html>